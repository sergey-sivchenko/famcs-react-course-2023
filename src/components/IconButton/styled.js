import styled from "styled-components/macro";

const PADDING_BY_SIZE = {
  small: 6,
  medium: 8,
  large: 10,
};

const getIconColor = (color, theme) => {
  switch (color) {
    case "primary": {
      return theme.primary.light;
    }

    case "secondary": {
      return theme.secondary.light;
    }

    case "primaryContrast": {
      return theme.primary.contrastText;
    }

    case "secondaryContrast": {
      return theme.secondary.contrastText;
    }

    default: {
      return theme.text.primary;
    }
  }
};

export const Container = styled.button`
  align-items: center;
  appearance: none;
  color: ${({ color, theme }) => getIconColor(color, theme)};
  background: transparent;
  border-radius: 50%;
  border: 0;
  cursor: pointer;
  display: flex;
  flex: 0;
  justify-content: center;
  margin: 0;
  outline: 0;
  padding: ${({ size }) => PADDING_BY_SIZE[size] || 8}px;
  text-align: center;
  text-decoration: none;
  user-select: none;
  vertical-align: middle;

  :hover {
    background-color: ${({ theme }) => theme.action.hover};
  }
`;
