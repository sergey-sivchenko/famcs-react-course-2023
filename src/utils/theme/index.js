import * as darkTheme from "./dark";
import * as lightTheme from "./light";

const themes = {
  dark: darkTheme,
  light: lightTheme,
};

export const createTheme = (mode = "dark") => ({
  action: themes[mode].action,
  background: themes[mode].background,
  common: {
    black: "#000000",
    white: "#ffffff",
  },
  divider: themes[mode].divider,
  mode,
  primary: themes[mode].primary,
  secondary: themes[mode].secondary,
  text: themes[mode].text,
});
