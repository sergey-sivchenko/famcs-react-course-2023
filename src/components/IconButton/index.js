import { Container } from "./styled";

const ICON_SIZES = {
  small: 24,
  medium: 28,
  large: 34,
};

const IconButton = ({
  color = "text",
  disabled = false,
  icon: Icon,
  label = "Icon button",
  onClick,
  size = "medium",
}) => (
  <Container aria-label={label} color={color} disabled={disabled} onClick={onClick} size={size}>
    <Icon size={ICON_SIZES[size] || 32} title={label} />
  </Container>
);

export default IconButton;
