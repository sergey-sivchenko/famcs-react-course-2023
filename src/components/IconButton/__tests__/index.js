import userEvent from "@testing-library/user-event";

import { render, screen } from "utils/testing";

import IconButton from "..";

describe("<IconButton/>", () => {
  it("matches snapshot with required props only", () => {
    const onClick = jest.fn();
    const Icon = jest.fn(({ size, title }) => `Icon: size - ${size}, title - ${title}`);

    const { asFragment } = render(<IconButton icon={Icon} onClick={onClick} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("calls onClick by a component click", async () => {
    const onClick = jest.fn();
    const Icon = jest.fn(({ size, title }) => `Icon: size - ${size}, title - ${title}`);

    render(<IconButton icon={Icon} onClick={onClick} />);

    await userEvent.click(screen.getByRole("button"));

    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
