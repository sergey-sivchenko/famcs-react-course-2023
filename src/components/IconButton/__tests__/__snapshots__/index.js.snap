// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton/> matches snapshot with required props only 1`] = `
<DocumentFragment>
  .c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  color: #f6f6f6;
  background: transparent;
  border-radius: 50%;
  border: 0;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 0;
  outline: 0;
  padding: 8px;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
}

.c0:hover {
  background-color: rgba(255 255 255 / 0.08);
}

<button
    aria-label="Icon button"
    class="c0"
    color="text"
  >
    Icon: size - 28, title - Icon button
  </button>
</DocumentFragment>
`;
